### PR TITLE
fix: renamed modulus to modulo; updated modulo operator defintion

### PR DESCRIPTION
**Pull request by jordanvieler**
BREAKING CHANGE: Renamed modulus to modulo. 

Added options and documentation for the modulo operator as defined in math and comp sci.

Refs: #353

### CODE REVIEW

**Comment by westonpace** [comment]
This is very helpful information, thank you for adding it.  I think this is all valid for modulo.  I have a question though.  Do you know of any query engines today that support more than just `truncate` behavior?

**Comment by richtia** [comment]
> This is very helpful information, thank you for adding it. I think this is all valid for modulo. I have a question though. Do you know of any query engines today that support more than just `truncate` behavior?

I was wondering the same.  Not exactly query engines, but this shows what different programming languages do: https://en.wikipedia.org/wiki/Modulo#In_programming_languages

**Review by richtia — APPROVED** [review]

**Comment by richtia** [comment]
Since this is a breaking change, we'll need 2 SMC approvals

**Comment by jordanvieler** [comment]
Per the Wikipedia link, I believe that ANSI SQL uses truncate division. So, presumably, any query engine compliant with the specification would use Truncate. However, the Python % operator utilizes Floored division, while the Rust % Operator utilizes Truncated division. Therefore, care would need to be taken to ensure that a Substrait producer written in Python and a Substrait consumer written in Rust consistently perform the Modulo operation.
This table from [Division and Modulus for Computer Scientists by Daan Leijen](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/divmodnote-letter.pdf) illustrates the differences in results from Truncate (T), Floor(F), and Euclidean (E) division. q is the quotient, and r is the remainder.

modulo_table

**Comment by westonpace** [comment]
Thanks for the example cases.

Rust, datafusion, sqlite, postgres, sql server, mysql, spark, and velox all use truncate.
Python, pandas, and cudf are the only examples I can find that use floor, but those are probably enough to justify it.

Let's update the PR to limit the choices to truncate and floor.  Given there are no systems that use ceil/round/euclidian I think the extra information would be confusing to users.

File: extensions/functions_arithmetic.yaml [hunk 1]
```diff
@@ -284,32 +284,89 @@ scalar_functions:
...
       - args:
           - name: x
             value: i8
           - name: y
             value: i8
+        options:
+          quotient:
+            values: [ TRUNCATE, FLOOR, CEILING, ROUND, EUCLIDEAN ]
+          overflow:
+            values: [ SILENT, SATURATE, ERROR ]
```
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:328]: Can we get an overflow performing modulo on integers if we are only using truncate and floor?

File: extensions/functions_arithmetic.yaml [hunk 2]
```diff
@@ -284,32 +284,89 @@ scalar_functions:
...
+          on_domain_error:
+            values: [ NAN, ERROR ]
```
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:330]: This implementation is for `i8` and `on_domain_error` does not make sense here.
```suggestion
```

File: extensions/functions_arithmetic.yaml [hunk 3]
```diff
@@ -284,32 +284,89 @@ scalar_functions:
...
             value: i16
           - name: y
             value: i16
+        options:
+          quotient:
+            values: [ TRUNCATE, FLOOR, CEILING, ROUND, EUCLIDEAN ]
+          overflow:
+            values: [ SILENT, SATURATE, ERROR ]
+          on_domain_error:
+            values: [ NAN, ERROR ]
```
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:343]:
```suggestion
```

File: extensions/functions_arithmetic.yaml [hunk 4]
```diff
@@ -284,32 +284,89 @@ scalar_functions:
...
             value: i32
           - name: y
             value: i32
+        options:
+          quotient:
+            values: [ TRUNCATE, FLOOR, CEILING, ROUND, EUCLIDEAN ]
+          overflow:
+            values: [ SILENT, SATURATE, ERROR ]
+          on_domain_error:
+            values: [ NAN, ERROR ]
```
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:356]:
```suggestion
```

File: extensions/functions_arithmetic.yaml [hunk 5]
```diff
@@ -284,32 +284,89 @@ scalar_functions:
...
             value: i64
           - name: y
             value: i64
+        options:
+          quotient:
+            values: [ TRUNCATE, FLOOR, CEILING, ROUND, EUCLIDEAN ]
+          overflow:
+            values: [ SILENT, SATURATE, ERROR ]
+          on_domain_error:
+            values: [ NAN, ERROR ]
```
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:369]:
```suggestion
```

**Review by westonpace — CHANGES_REQUESTED** [review]
If we want `on_domain_error` then we will need to add floating point implementations.  Also see my change from my non-review comment about removing `ceiling`, `round`, and `euclidian`.

**Comment by jordanvieler** [comment]
Thanks for the feedback. I just have a couple questions.

1) How is on_domain_error meant to be defined in substrait? The Modulo function is undefined (return ∅) when the divisor is 0 or arguments approach +/- inf. It is my understanding this is "out of domain." Therefore, it is out of domain. How relevant is float vs integer division in this case?

2) While python uses floor for its % operator in the example, most languages define a set of functions in their standard library. For example, Rust defines rem_euclid() for the Euclidean definition of division. Swift defines remainder(dividingBy:) for a modulo with a rounded definition.

Because of how drastic the results can be when performing these different types of modulo, I think the distinctions are important for Data Science, Statistics, and Machine Learning applications. For example a Data Scientist might specifically call for a Ceil Modulo to be performed from their python data frame library.

File: extensions/functions_arithmetic.yaml (hunk 1 above)
Comment (jordanvieler) [reply extensions/functions_arithmetic.yaml:328]: Correct me if I am wrong, but I think all modulo can overflow if the dividend is equal to the minimum negative value for a signed int and the divisor is equal to -1.

File: extensions/functions_arithmetic.yaml (hunk 2 above)
Comment (jordanvieler) [reply extensions/functions_arithmetic.yaml:330]: I think modulo is undefined for a divisor of 0 and args of +/- inf.

**Comment by westonpace** [comment]
> How is on_domain_error meant to be defined in substrait? The Modulo function is undefined (return ∅) when the divisor is 0 or arguments approach +/- inf. Therefore, I believe it is "out of domain" by definition. How relevant is float vs integer division in this case?

Float vs integer is relevant because different implementations (e.g f32 vs i32) have different options.  Historically we've used `overflow=SILENT, SATURATE, ERROR` for integer operations and `on_domain_error=NAN, ERROR` for floating point operations.  We could have done this other ways too.  For example, integer division could have both `overflow=SILENT,SATURATE,ERROR` (to handle `min_int/-1`) and `on_domain_error=SILENT,ERROR` (for division by zero).  We went with the simplest approach to match what was currently being offered by available engines.

The current PR's proposed `on_domain_error` will not work for integer implementations however because `NAN` is not possible.

This does answer my "is overflow possible for modulo?" question though (if we consider division by zero to be overflow).

> While python uses floor for its % operator in the example, many languages also define a set of other modulo functions in their libraries. For example, Rust defines rem_euclid() for the Euclidean definition of division. Swift defines remainder(dividingBy:) for a modulo with a rounded definition. Are these important cases?

Our current charter for standard functions is to only capture functions that a sufficient number of query engines currently provide.  Rust may offer `rem_euclid` but datafusion does not.  Even if one engine were to provide such a method that would be a better case for an engine-specific extension function than a standard Substrait function.

**Comment by jordanvieler** [comment]
> Float vs integer is relevant because different implementations (e.g f32 vs i32) have different options. Historically we've used `overflow=SILENT, SATURATE, ERROR` for integer operations and `on_domain_error=NAN, ERROR` for floating point operations. We could have done this other ways too. For example, integer division could have both `overflow=SILENT,SATURATE,ERROR` (to handle `min_int/-1`) and `on_domain_error=SILENT,ERROR` (for division by zero). We went with the simplest approach to match what was currently being offered by available engines.
> 
> The current PR's proposed `on_domain_error` will not work for integer implementations however because `NAN` is not possible.
> 
> This does answer my "is overflow possible for modulo?" question though (if we consider division by zero to be overflow).
> 
Ok, I see what you are saying about NaN not being relevant for int types. Thank you for clarifying that. However, I think there is a difference between overflow and out-of-domain for modulo. Do we assume that a Substrait consumer will return an Error for out-of-domain values of the modulus function, namely 0, as a divisor? So, no option is necessary, right? I could also imagine a historical argument for MAX_INT being a return value in these cases.

> 
> Our current charter for standard functions is to only capture functions that a sufficient number of query engines currently provide. Rust may offer `rem_euclid` but datafusion does not. Even if one engine were to provide such a method that would be a better case for an engine-specific extension function than a standard Substrait function.

I understand the rationale, but I am approaching this slightly differently. Shouldn't standard Substrait functions be toward creating a platform, language, and system agnostic specification and network protocol for data computing operations? I think only supporting the current behavior of existing query engines could introduce a historical bias and limit use cases. SQL itself is only loosely based on relational algebra.

**Comment by westonpace** [comment]
> Do we assume that a Substrait consumer will return an Error for out-of-domain values of the modulus function, namely 0, as a divisor? So, no option is necessary, right?

Sadly, it seems we are not so lucky.

Postgres, Datafusion, SQL server, velox raise an error
Sqlite, spark, and MySQL return null
Pandas and cudf silently upcast the entire operation to floating point and return NaN (this would not be valid substrait because the return type is not allowed to depend on input and so we can probably ignore this case)

Even ignoring the invalid case we still have to consider error and null options.

> I understand the rationale, but I am approaching this slightly differently. Shouldn't standard Substrait functions be toward creating a platform, language, and system agnostic specification and network protocol for data compute operations? I think only supporting the current behavior of existing query engines could introduce a historical bias and limit use cases. SQL itself is only loosely based on relational algebra.

This is a fair viewpoint but the flipside is that anyone using Substrait to create plans cannot reasonably assume the plan will run.  Either way, this issue is probably not the place to discuss the charter for functions.  I'm afraid I'm rather inflexible.  I've been reviewing based on the last time we discussed this topic which was at https://github.com/substrait-io/substrait/issues/307.

If we want a new charter then we should have a separate discussion where everyone can weigh in.  Either as a new issue, in a community meeting, or on the mailing list.  Discussing the scope of Substrait functions on a per-function / per-PR basis doesn't really work.

**Comment by jordanvieler** [comment]
> Sadly, it seems we are not so lucky.
> 
> Postgres, Datafusion, SQL server, velox raise an error Sqlite, spark, and MySQL return null Pandas and cudf silently upcast the entire operation to floating point and return NaN (this would not be valid substrait because the return type is not allowed to depend on input and so we can probably ignore this case)
> 
> Even ignoring the invalid case we still have to consider error and null options.
> 

Does this mean that a field to specify domain errors is warranted separately from the overflow error we have discussed?

> This is a fair viewpoint but the flipside is that anyone using Substrait to create plans cannot reasonably assume the plan will run. Either way, this issue is probably not the place to discuss the charter for functions. I'm afraid I'm rather inflexible. I've been reviewing based on the last time we discussed this topic which was at #307.

Per the referenced topic it seems like modulo is exactly the pedantic case that @jvanstraten mentions.

> If we want a new charter then we should have a separate discussion where everyone can weigh in. Either as a new issue, in a community meeting, or on the mailing list. Discussing the scope of Substrait functions on a per-function / per-PR basis doesn't really work.

I completely understand. I am relatively new to contributing to open-source. My attempt to resolve the divide-by-zero behavior highlighted these nuances modulo. Would it then make sense to bring this aspect of the discussion somewhere else?

**Comment by jordanvieler** [comment]
For now how does this look?

File: extensions/functions_arithmetic.yaml [hunk 6]
```diff
@@ -284,32 +284,84 @@ scalar_functions:
...
-    name: "modulus"
-    description: "Get the remainder when dividing one value by another."
+    name: "modulo"
+    description: >
+      Calculate the remainder (r) when dividing dividend (x) by divisor (y).
+
+      In mathematics, many conventions for the modulo operation exists. In computing the result of a modulo operation
+      depends on the software implementation and underlying hardware. Substrait is a format for describing compute
+      operations on structured data and designed for interoperability. Therefore the user is responsible for determining
+      a definition of division as defined by the quotient (y).
```
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:294]:
```suggestion
      a definition of division as defined by the quotient (q).
```

File: extensions/functions_arithmetic.yaml [hunk 7]
```diff
@@ -284,32 +284,84 @@ scalar_functions:
             value: fp64
         return: fp64
   -
-    name: "modulus"
-    description: "Get the remainder when dividing one value by another."
+    name: "modulo"
+    description: >
+      Calculate the remainder (r) when dividing dividend (x) by divisor (y).
+
+      In mathematics, many conventions for the modulo operation exists. In computing the result of a modulo operation
```
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:291]:
```suggestion
      In mathematics, many conventions for the modulo operation exists. The result of a modulo operation
```
This read a little odd to me.  I wasn't sure if that is "in the field of computing, the result..." or "while computing the result...".  I think simpler might be better here.

File: extensions/functions_arithmetic.yaml [hunk 8]
```diff
@@ -284,32 +284,84 @@ scalar_functions:
...
+
+      When `division_type`=TRUNCATE, q = trunc(x/y).
+      When `division_type`=FLOOR, q = floor(x/y).
+
+      In the cases of TRUNCATE and FLOOR division: remainder r = x - round_func(x/y)
+
+      The `on_domain_error` option governs behavior in cases where y is 0, y is +/-inf, or x is +/-inf. In these cases
+      modulo is undefined.
+      The `overflow` option governs behavior when integer overflow occurs.
+      If x and y are both 0 or both +/-infinity, behavior will be governed by `on_domain_error`.
```
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:313]: Integers can't represent infinity.  We can either drop all discussion of infinity for now or add `impls` for `fp32` and `fp64`.

File: extensions/functions_arithmetic.yaml [hunk 9]
```diff
@@ -284,32 +284,84 @@ scalar_functions:
...
           - name: x
             value: i8
           - name: y
             value: i8
+        options:
+          division_type:
+            values: [ TRUNCATE, FLOOR ]
+          overflow:
+            values: [ SILENT, SATURATE, ERROR ]
+          on_domain_error: values: [ NAN, ERROR ]
```
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:325]:
```suggestion
          on_domain_error: values: [ NULL, ERROR ]
```

File: extensions/functions_arithmetic.yaml [hunk 10]
```diff
@@ -284,32 +284,84 @@ scalar_functions:
...
             value: i16
           - name: y
             value: i16
+        options:
+          division_type:
+            values: [ TRUNCATE, FLOOR ]
+          overflow:
+            values: [ SILENT, SATURATE, ERROR ]
+          on_domain_error:
+            values: [ NAN, ERROR ]
```
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:338]:
```suggestion
            values: [ NULL, ERROR ]
```

File: extensions/functions_arithmetic.yaml [hunk 11]
```diff
@@ -284,32 +284,84 @@ scalar_functions:
...
             value: i32
           - name: y
             value: i32
+        options:
+          division_type:
+            values: [ TRUNCATE, FLOOR ]
+          overflow:
+            values: [ SILENT, SATURATE, ERROR ]
+          on_domain_error:
+            values: [ NAN, ERROR ]
```
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:351]:
```suggestion
            values: [ NULL, ERROR ]
```

File: extensions/functions_arithmetic.yaml [hunk 12]
```diff
@@ -284,32 +284,84 @@ scalar_functions:
...
             value: i64
           - name: y
             value: i64
+        options:
+          division_type:
+            values: [ TRUNCATE, FLOOR ]
+          overflow:
+            values: [ SILENT, SATURATE, ERROR ]
+          on_domain_error:
+            values: [ NAN, ERROR ]
```
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:364]:
```suggestion
            values: [ NULL, ERROR ]
```

**Review by westonpace — CHANGES_REQUESTED** [review]
Let's keep both `overflow` and `on_domain_error` but change `NAN` to `NULL` since we are dealing with integer kernels which cannot represent `NAN`.

File: extensions/functions_arithmetic.yaml (hunk 8 above)
Comment (jordanvieler) [reply extensions/functions_arithmetic.yaml:313]: Ah yes. Thanks for catching that

File: extensions/functions_arithmetic.yaml [hunk 13]
```diff
@@ -284,32 +284,82 @@ scalar_functions:
...
           - name: x
             value: i8
           - name: y
             value: i8
+        options:
+          division_type:
+            values: [ TRUNCATE, FLOOR ]
+          overflow:
+            values: [ SILENT, SATURATE, ERROR ]
+          on_domain_error: values: [ NULL, ERROR ]
```
Comment (EpsilonPrime) [inline extensions/functions_arithmetic.yaml:325]: Looks like this should be on two lines.

Comment (jordanvieler) [reply extensions/functions_arithmetic.yaml]: My Bad Thanks.

**Comment by jordanvieler** [comment]
It seems like NULL is not valid for the YAML Linter.

File: extensions/functions_arithmetic.yaml [hunk 14]
```diff
@@ -284,32 +284,83 @@ scalar_functions:
...
             value: i8
           - name: y
             value: i8
+        options:
+          division_type:
+            values: [ TRUNCATE, FLOOR ]
+          overflow:
+            values: [ SILENT, SATURATE, ERROR ]
+          on_domain_error:
+            values: [ NULL, ERROR ]
```
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:326]:
```suggestion
            values: [ "NULL", ERROR ]
```

File: extensions/functions_arithmetic.yaml [hunk 15]
```diff
@@ -284,32 +284,83 @@ scalar_functions:
...
             value: i32
           - name: y
             value: i32
+        options:
+          division_type:
+            values: [ TRUNCATE, FLOOR ]
+          overflow:
+            values: [ SILENT, SATURATE, ERROR ]
+          on_domain_error:
+            values: [ NULL, ERROR ]
```
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:350]:
```suggestion
            values: [ "NULL", ERROR ]
```

File: extensions/functions_arithmetic.yaml [hunk 16]
```diff
@@ -284,32 +284,83 @@ scalar_functions:
...
             value: i64
           - name: y
             value: i64
+        options:
+          division_type:
+            values: [ TRUNCATE, FLOOR ]
+          overflow:
+            values: [ SILENT, SATURATE, ERROR ]
+          on_domain_error:
+            values: [ NULL, ERROR ]
```
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:363]:
```suggestion
            values: [ "NULL", ERROR ]
```

**Review by westonpace — CHANGES_REQUESTED** [review]
Seems to be an annoying yaml / yaml-validator gotcha.  `NULL` is being interpreted as the data type "null" and not as a string.

**Comment by westonpace** [comment]
The one thing we haven't talked about is whether the name change justifies breaking backwards compatibility.  We have a community meeting coming up on Wednesday and I think we should bring this up there.

**Comment by jordanvieler** [comment]
Is it necessary? I added that change after making a very deep dive into mod and thought I read modulus referred to something else. But now I think it's just preference.

**Comment by westonpace** [comment]
> Is it necessary? I added that change after making a very deep dive into mod and thought I read modulus referred to something else. But now I think it's just preference.

If you're comfortable with the old name then let's keep it.  There are systems out there which are relying on the old name and so the breaking change would be a bit inconvenient.

**Review by westonpace — APPROVED** [review]
This looks good to me now.  Thanks for all your work!  @EpsilonPrime did you want to take another look?

**Review by EpsilonPrime — APPROVED** [review]

**Pull request merged**